feat($goSync): Sync local object w. GoInstant key

mattcreager opened the pull request:
The $goKey service has been introduced, it takes a single argument, a
key.  $goKey returns a model, the standard `key` methods are exposed,
prefixed with a '$'. This model will be kept current with the key.

Note: Changes to the object are not persisted back to the key.

For example, to return a collection of todos:

``` js
$scope.todos = $goKey(lobby.key('todos'));
$scope.todos.$sync();
```

To add a new todo:

``` js
$scope.todos.$add({
  description: 'a fresh todo',
  complete: false
});
```

Each of the methods with the exception of $sync, returns a promise.

Additional minor changes:

Also added grunt-contrib-watch, it adds `grunt watch`, which:
- monitors the lib folder and runs component build -d
- monitors the build file and live-reloads

Closes #28, #23, #31

ianlivingstone commented on lib/model_binder.js:
```diff
@@ -99,11 +100,16 @@ ModelBinder.prototype._firstFetch = function(model, next) {
     var hasDefault = (!_.isUndefined(modelDefault) && !_.isNull(modelDefault));
     var hasValue = (!_.isUndefined(value) && !_.isNull(value));
 
+    console.log(value)
```
dont commit console logs

Commit:
feat($goKey): Sync local object w. GoInstant key

The $goKey service has been introduced, it takes a single argument, a
key.  $goKey returns a model, the standard `key` methods are exposed,
prefixed with a '$'. This model will be kept current with the key.

Note: Changes to the object are not persisted back to the key.

For example, to return a collection of todos:

$scope.todos = $goKey(lobby.key('todos'));
$scope.todos.$sync();

To add a new todo:

$scope.todos.$add({
  description: 'a fresh todo',
  complete: false
});

Each of the methods with the exception of $sync, returns a promise.

Additional minor changes:

Also added grunt-contrib-watch, it adds `grunt watch`, which:
- monitors the lib folder and runs component build -d
- monitors the build file and live-reloads

Closes #28, #23

Commit:
docs(model.md): Model landing page / guide

Commit:
docs(docs/v1/model/*): Adding GoAngular model documentation

ianlivingstone commented on Gruntfile.js:
```diff
@@ -37,6 +37,7 @@ module.exports = function(grunt) {
   grunt.loadNpmTasks('grunt-contrib-clean');
   grunt.loadNpmTasks('grunt-contrib-connect');
   grunt.loadNpmTasks('grunt-saucelabs');
+  grunt.loadNpmTasks('grunt-contrib-watch');
```
Changes like this should be done in a separate pull request into master in future.

mattcreager replied:
https://github.com/goinstant/goangular/pull/29  <--- It was, but I needed the watch task to work efficiently!

ianlivingstone replied:
Hmm, why is it showing up in the diff then

ianlivingstone commented on component.json:
```diff
@@ -7,7 +7,8 @@
     "caolan/async": "c99537ae4837f12e918959c8bd1e70e6b1730f6e",
-    "lodash/lodash": "2.1.0"
+    "lodash/lodash": "2.1.0",
+    "microjs/q": "*"
```
set a specific version please

mattcreager replied:
removed, was just in for testing!

ianlivingstone commented on lib/connection.js:
```diff
@@ -59,10 +64,6 @@ Connection.prototype.set = function(url, opts) {
  * the connection with GoInstant
  */
 Connection.prototype.$get = function() {
-  if (!window.goinstant) {
```
why'd you remove it here but add it to `set` ?

mattcreager replied:
Set happens first, just erroring as early as possible.

ianlivingstone replied:
for completeness both should have the check

mattcreager replied:
Just to confirm, $get cannot be called without $set being called first, it validates the connection is configured and throws otherwise... so $set will always be called... still worth the second check?

mattcreager commented on lib/key_factory.js:
```diff
@@ -0,0 +1,38 @@
+   * @public
+   * @param {Object} key - GoInstant key
+   */
+  return function(key, room) {
```
just added the room parameter, needs validation.

mattcreager commented on lib/key_factory.js:
```diff
@@ -0,0 +1,38 @@
+   */
+  return function(key, room) {
+
+    if (!key) {
```
Can be a string or an object now.

mattcreager commented on lib/model.js:
```diff
@@ -0,0 +1,149 @@
+ * @param {Object} $goConnection - GoInstant connection service
+ * @param {Object} key - GoInstant Key
+ * @example
+ *   $scope.todos = $goSync(lobby.key('todos')).$get();
```
dated, I've got to update this.

ianlivingstone commented on lib/model_binder.js:
```diff
@@ -57,7 +58,7 @@ ModelBinder.prototype.newBinding = function(modelName, cb) {
   );
 
-  async.parallel(boundFuncs, cb);
+  async.series(boundFuncs, cb);
```
why'd this have to change?

mattcreager replied:
this is getting cut, it's been pulled out of the documentation.

ianlivingstone commented on lib/sync.js:
```diff
@@ -0,0 +1,207 @@
+      _.isString,
+      _.isNumber,
+      _.isBoolean,
+      _.isDate,
```
a date isn't a primitive, its easy just to check its not an object. This is doing a lot of work which could just be `if (!_.isObject())`

mattcreager replied:
The problem is, a date is an object, but should be stored at $value

ianlivingstone replied:
That should be dealt as a special case where you call .toString() since Platform does not handle `Date` objects.

mattcreager replied:
Ok I'm going to validate against the object for now, and handle the special case of a date in a separate PR..

ianlivingstone commented on lib/sync.js:
```diff
@@ -0,0 +1,207 @@
+    if (isPrimitive) {
+      model = { $value: value };
+    } else {
+      _.each(value, function(value, key) {
```
whats the difference between this and a `_.merge` ?

mattcreager replied:
...... All I can say is, I wrote this during Christmas break... I must have had too much eggnog.

Commit:
Addressing CR feedback

Commit:
feat($goKey): Accepts a string vs. a key

goKey will now accept a key, string or an optional room name, defaulting
to 'lobby'.  It will ensure that the room has been joined prior to
perfoming an operation on a key.

resolves #31

Conflicts:
	docs/v1/API/goConnection.md

mattcreager commented:
Verbal LGTM received via. @ianlivingstone
